### DISCUSSION TITLE
Using a struct with HSet

### DISCUSSION THREAD

**Discussion by xorduna** [redis/go-redis#2454] [Q&A] [answered]
Hi!

According to the documentation, I should be able to perform an HSet Command using an struct.

In the documentation I found this example:

```
Playing struct With "redis" tag. type MyHash struct { Key1 string `redis:"key1"`; Key2 int `redis:"key2"` }

HSet("myhash", MyHash{"value1", "value2"})

For struct, can be a structure pointer type, we only parse the field whose tag is redis. if you don't want the field to be read, you can use the `redis:"-"` flag to ignore it, or you don't need to set the redis tag. For the type of structure field, we only support simple data types: string, int/uint(8,16,32,64), float(32,64), time.Time(to RFC3339Nano), time.Duration(to Nanoseconds ), if you are other more complex or custom data types, please implement the encoding.BinaryMarshaler interface. 
```

When when trying to run the example

```golang
type Info struct {
	Testgroup int    `json:"testgroup"  redis:"testgroup"`
	Status    string `json:"status" redis:"status"`
}

func main() {
	rdb := redis.NewClient(&redis.Options{
		Addr: ":6379",
	})

	ctx := context.Background()
	cmd := rdb.HSet(context.Background(), "mysupermap", models.Info{6, "RUNNING"})
	i, err := cmd.Result()
	fmt.Println(i, err)
}
```
I am getting this error: ERR wrong number of arguments for 'hset' command

What am I doing wrong?

Thank you very much!!

**Comment by monkey92t** [comment]
I ran your example and it worked fine, the version I'm using is redis/go-redis/v9@v9.0.2

**Comment by monkey92t** [comment]
my go version: go1.20, redis-version: 7.0.8, go-redis version: v9.0.2

```go
type Per struct {
	Name string `redis:"name"`
	Age  int    `redis:"age"`
}

func main() {
	client := redis.NewClient(&redis.Options{
		Addr: "127.0.0.1:6379",
	})
	ctx := context.Background()

	err := client.HSet(ctx, "myhash", Per{Name: "hi", Age: 20}).Err()
	fmt.Println(err) // print: nil
}

127.0.0.1:6379> hgetall myhash
1) "name"
2) "hi"
3) "age"
4) "20"
127.0.0.1:6379>
```

**Reply by monkey92t** [reply]
Before version 4.0 of Redis, the hset command only supported a single key-value pair. If you are using an earlier version of Redis, you will receive a `ERR wrong number of arguments` error in response from the Redis server.

Later on, I will add comments to the HSET API to help more people avoid this issue.

**Comment by monkey92t** [comment, marked as answer]
Note that in older versions of Redis server(redis-server < 4.0), HSet only supports a single key-value pair.

redis-docs: https://redis.io/commands/hset (Starting with Redis version 4.0.0: Accepts multiple field and value arguments.)

If you are using a Struct type and the number of fields is greater than one, you will receive an error similar to `ERR wrong number of arguments`, you can use HMSet as a substitute.

redis-server >= 4.0(HSet):
```go
client.HSet(ctx, "myhash", Per{Name: "hi", Age: 20})
```

redis-server < 4.0(HMSet):
```go
client.HMSet(ctx, "myhash", Per{Name: "hi", Age: 20})
```

Please note that in Redis, HMSet has been marked as `deprecated`, and it may not be supported in the future.

#2503

**Comment by 0x30c4** [comment]
hello! @monkey92t. 
this code doesn't work 
```go
client.HSet(context.Background(), "myhash", Per{Name: "hi", Age: 20})
```
```
cannot use context.Background() (value of type context.Context) as string value in argument to client.HSet```

**Comment by monkey92t** [comment]
> hello! @monkey92t. this code doesn't work
> 
> ```go
> client.HSet(context.Background(), "myhash", Per{Name: "hi", Age: 20})
> ```
> 
> ```
> cannot use context.Background() (value of type context.Context) as string value in argument to client.HSet```
> ```

You seem to be using a lower version of go-redis. Could you please provide your redis-server version and go-redis version?

**Reply by 0x30c4** [reply]
yeah, i was indeed using go-redis v6.
thank you.

**Comment by albertosml** [comment]
Note that I also get the `ERR wrong number of arguments for 'hset' command` error for maps as well. I can put an example:

```go
client.HSet(ctx, "myhash", map[string]interface{}{"name": "hi", "age": 20})
```

However, I don't experience it when my map is only composed by one field.

The issue has been found on version 8.11.5.
